File: Kalorie/Kalorie/Core/Models/FoodPortionModel.swift
//
//  FoodPortionModel.swift
//  Kalorie
//
//  Created by Josef Antoni on 09.09.2026.
//

import Foundation

struct FoodPortionDomain: Hashable {

    // MARK: - Properties

    let name: String
    let grams: Double
}

enum FoodPortionError: Error {
    case invalidName
    case invalidGrams
    case tooMany
}

enum FoodPortionValidation {

    // MARK: - Properties

    static let maxPortions = 20

    // MARK: - Functions

    static func validate(name: String, grams: Double) -> FoodPortionError? {
        guard !name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return .invalidName }
        guard grams >= 1 else { return .invalidGrams }
        return nil
    }

    static func validate(portions: [FoodPortionDomain]) -> FoodPortionError? {
        portions.count > maxPortions ? .tooMany : nil
    }
}

File: Android/app/src/main/java/antoni/kalorie/core/models/FoodPortionModel.kt
package antoni.kalorie.core.models

import androidx.annotation.StringRes
import antoni.kalorie.R

data class FoodPortionDomain(
    val name: String,
    val grams: Double,
)

sealed class FoodPortionError : Exception() {
    data object InvalidName : FoodPortionError()
    data object InvalidGrams : FoodPortionError()
    data object TooMany : FoodPortionError()

    // MARK: - Properties

    @get:StringRes
    val alertTitleRes: Int
        get() = when (this) {
            InvalidName -> R.string.foodPortion_error_invalidName
            InvalidGrams -> R.string.foodPortion_error_invalidGrams
            TooMany -> R.string.foodPortion_error_tooMany
        }
}

object FoodPortionValidation {

    // MARK: - Properties

    const val MAX_PORTIONS = 20

    // MARK: - Functions

    fun validate(name: String, grams: Double): FoodPortionError? {
        if (name.isBlank()) return FoodPortionError.InvalidName
        if (grams < 1) return FoodPortionError.InvalidGrams
        return null
    }

    fun validate(portions: List<FoodPortionDomain>): FoodPortionError? =
        if (portions.size > MAX_PORTIONS) FoodPortionError.TooMany else null
}
